src/Filters/RelatedDto.php

```diff
@@ -83,56 +67,146 @@ public function isResolved(string $relationship): bool
+            return $relatedQuery;
+        }
 
-            $this->loaded = true;
+        if ($relatedQuery->nested->count()) {
+            return $relatedQuery->nested->first(fn (RelatedQuery $child) => $this->searchInRelatedQuery(
+                $child,
+                $relation
+            ));
         }
 
-        return $this;
+        return null;
     }
 
-    public function reset(): self
+    private function makeTreeForChild(RelatedQuery $relatedQuery, array &$base = []): array
     {
-        $this->loaded = false;
+        if ($relatedQuery->nested->count()) {
+            $relatedQuery->nested->each(function (RelatedQuery $child, $i) use (&$base, $relatedQuery) {
+                $base[$i] = data_get($base, $i, $relatedQuery->relation).".$child->relation";
+
+                if ($child->nested->count()) {
+                    $this->makeTreeForChild($child, $base);
+                }
+            });
+        } else {
+            return [$relatedQuery->relation];
+        }
 
-        $this->resolvedRelationships = [];
+        return $base;
+    }
 
-        return $this;
+    public function makeTree(): array
+    {
+        return collect(array_keys($this->relatedArray))
+            ->mapInto(Stringable::class)
+            ->map(fn (Stringable $relation) => $relation->after('.'))
+            ->unique()
+            ->map(fn (Stringable $relation) => $relation->toString())
+            ->all();
     }
 
-    private function makeTreeFor(string $related, ?RelatedDto $dto = null): string
+    public function sync(Request $request, Repository $repository): self
     {
-        if (is_null($dto)) {
-            return $related;
+        if ($this->loaded) {
+            return $this;
+        }
+
+        if (empty($query = $this->query($request))) {
+            $this->loaded();
+
+            return $this;
         }
 
-        $child = collect($dto->related)->first();
+        $roots = str($query)->replace(' ', '')->explode(',');
+
+        collect($roots)->map(function (string $related) use ($repository) {
+            if (str($related)->contains('.')) {
+                // users[id].comments[id] => users
+                $relation = str(collect(str($related)->explode('.'))->first())->before('[');
+            } else {
+                // comments[id] => comments
+                // comments => comments
+                $relation = str($related)->before('[');
+            }
+
+            /**
+             * @var RelatedQuery|null $relatedQuery
+             */
+            if ($relatedQuery = $this->related->firstWhere('relation', $relation)) {
+                $parent = $relatedQuery;
+            } else {
+                $parent = RelatedQueryCollection::fromToken(str($related)->before('.'))->parent($repository::uriKey());
+                $this->relatedArray[$parent->tree] = clone $parent;
+            }
+
+            // Here it's like `comments[id]`
+            if (! str($related)->contains('.')) {
```

Comment:
Or, you could save the `str($related)` to a variable and not create as many objects :P